File: frontend/src/services/blogService.js
import api from './api.js'

export const blogService = {
    // Blog operations
    async getAllBlogs(page = 0, size = 10) {
        try {
            const response = await api.get(`/blogs?page=${page}&size=${size}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch blogs'
            }
        }
    },

    async getAllBlogsList() {
        try {
            const response = await api.get('/blogs/list')
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch blogs'
            }
        }
    },

    async getBlogById(id) {
        try {
            const response = await api.get(`/blogs/${id}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch blog'
            }
        }
    },

    async createBlog(blogData) {
        try {
            const response = await api.post('/blogs', blogData)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to create blog'
            }
        }
    },

    async updateBlog(id, blogData) {
        try {
            const response = await api.put(`/blogs/${id}`, blogData)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to update blog'
            }
        }
    },

    async deleteBlog(id) {
        try {
            const response = await api.delete(`/blogs/${id}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to delete blog'
            }
        }
    },

    async publishBlog(id) {
        try {
            const response = await api.put(`/blogs/${id}/publish`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to publish blog'
            }
        }
    },

    async unpublishBlog(id) {
        try {
            const response = await api.put(`/blogs/${id}/unpublish`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to unpublish blog'
            }
        }
    },

    async getBlogsByAuthor(authorId, page = 0, size = 10, includeUnpublished = false) {
        try {
            const response = await api.get(`/blogs/author/${authorId}?page=${page}&size=${size}&includeUnpublished=${includeUnpublished}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch author blogs'
            }
        }
    },

    async searchBlogs(keyword) {
        try {
            const response = await api.get(`/blogs/search?keyword=${encodeURIComponent(keyword)}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to search blogs'
            }
        }
    },

    async getRecentBlogs(days = 7) {
        try {
            const response = await api.get(`/blogs/recent?days=${days}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch recent blogs'
            }
        }
    },

    async getPopularBlogs(limit = 10) {
        try {
            const response = await api.get(`/blogs/popular?limit=${limit}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch popular blogs'
            }
        }
    },

    async getBlogStats() {
        try {
            const response = await api.get('/blogs/stats')
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch blog stats'
            }
        }
    },

    // Comment operations
    async getCommentsByBlog(blogId, page = 0, size = 20) {
        try {
            const response = await api.get(`/comments/blog/${blogId}?page=${page}&size=${size}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch comments'
            }
        }
    },

    async getCommentsByBlogList(blogId) {
        try {
            const response = await api.get(`/comments/blog/${blogId}/list`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch comments'
            }
        }
    },

    async getCommentsByAuthor(authorId) {
        try {
            const response = await api.get(`/comments/author/${authorId}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch author comments'
            }
        }
    },

    async createComment(commentData) {
        try {
            const response = await api.post('/comments', commentData)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to create comment'
            }
        }
    },

    async createCommentForBlog(blogId, content, authorId) {
        try {
            const response = await api.post(`/comments/blog/${blogId}`, {
                content,
                authorId
            })
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to create comment'
            }
        }
    },

    async updateComment(id, content) {
        try {
            const response = await api.put(`/comments/${id}`, { content })
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to update comment'
            }
        }
    },

    async deleteComment(id) {
        try {
            const response = await api.delete(`/comments/${id}`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to delete comment'
            }
        }
    },

    async getCommentCount(blogId) {
        try {
            const response = await api.get(`/comments/blog/${blogId}/count`)
            return {
                success: true,
                data: response.data
            }
        } catch (error) {
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to fetch comment count'
            }
        }
    }
}

File: src/main/java/com/turningpoint/chapterorganizer/service/BlogService.java
package com.turningpoint.chapterorganizer.service;

import com.turningpoint.chapterorganizer.entity.Blog;
import com.turningpoint.chapterorganizer.entity.Comment;
import com.turningpoint.chapterorganizer.repository.BlogRepository;
import com.turningpoint.chapterorganizer.repository.CommentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class BlogService {

    private final BlogRepository blogRepository;
    private final CommentRepository commentRepository;

    @Autowired
    public BlogService(BlogRepository blogRepository, CommentRepository commentRepository) {
        this.blogRepository = blogRepository;
        this.commentRepository = commentRepository;
    }

    /**
     * Get all published blogs
     */
    @Transactional(readOnly = true)
    public List<Blog> getAllPublishedBlogs() {
        return blogRepository.findByPublishedTrueOrderByCreatedAtDesc();
    }

    /**
     * Get all blogs (including drafts)
     */
    @Transactional(readOnly = true)
    public List<Blog> getAllBlogs() {
        return blogRepository.findAllByOrderByCreatedAtDesc();
    }

    /**
     * Get blog by ID
     */
    @Transactional(readOnly = true)
    public Optional<Blog> getBlogById(Long id) {
        return blogRepository.findById(id);
    }

    /**
     * Get blogs by author
     */
    @Transactional(readOnly = true)
    public List<Blog> getBlogsByAuthor(Long authorId) {
        return blogRepository.findByAuthorIdOrderByCreatedAtDesc(authorId);
    }

    /**
     * Get published blogs by author
     */
    @Transactional(readOnly = true)
    public List<Blog> getPublishedBlogsByAuthor(Long authorId) {
        return blogRepository.findByAuthorIdAndPublishedTrueOrderByCreatedAtDesc(authorId);
    }

    /**
     * Create new blog
     */
    public Blog createBlog(Blog blog) {
        if (blog.getAuthor() == null) {
            throw new IllegalArgumentException("Blog must have an author");
        }
        if (blog.getTitle() == null || blog.getTitle().trim().isEmpty()) {
            throw new IllegalArgumentException("Blog title is required");
        }
        if (blog.getContent() == null || blog.getContent().trim().isEmpty()) {
            throw new IllegalArgumentException("Blog content is required");
        }
        
        // Set default published status if not provided
        if (blog.getPublished() == null) {
            blog.setPublished(false);
        }
        
        return blogRepository.save(blog);
    }

    /**
     * Update blog
     */
    public Blog updateBlog(Long id, Blog blogUpdate) {
        Blog existingBlog = blogRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Blog not found with id: " + id));

        // Update fields
        if (blogUpdate.getTitle() != null && !blogUpdate.getTitle().trim().isEmpty()) {
            existingBlog.setTitle(blogUpdate.getTitle());
        }
        if (blogUpdate.getContent() != null && !blogUpdate.getContent().trim().isEmpty()) {
            existingBlog.setContent(blogUpdate.getContent());
        }
        if (blogUpdate.getPublished() != null) {
            existingBlog.setPublished(blogUpdate.getPublished());
        }

        return blogRepository.save(existingBlog);
    }

    /**
     * Publish blog
     */
    public Blog publishBlog(Long id) {
        Blog blog = blogRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Blog not found with id: " + id));
        
        blog.setPublished(true);
        return blogRepository.save(blog);
    }

    /**
     * Unpublish blog (make it a draft)
     */
    public Blog unpublishBlog(Long id) {
        Blog blog = blogRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Blog not found with id: " + id));
        
        blog.setPublished(false);
        return blogRepository.save(blog);
    }

    /**
     * Delete blog
     */
    public void deleteBlog(Long id) {
        if (!blogRepository.existsById(id)) {
            throw new IllegalArgumentException("Blog not found with id: " + id);
        }
        blogRepository.deleteById(id);
    }

    /**
     * Add comment to blog
     */
    public Comment addComment(Long blogId, Comment comment) {
        Blog blog = blogRepository.findById(blogId)
                .orElseThrow(() -> new IllegalArgumentException("Blog not found with id: " + blogId));

        if (comment.getAuthor() == null) {
            throw new IllegalArgumentException("Comment must have an author");
        }
        if (comment.getContent() == null || comment.getContent().trim().isEmpty()) {
            throw new IllegalArgumentException("Comment content is required");
        }

        comment.setBlog(blog);
        Comment savedComment = commentRepository.save(comment);
        
        // Add to blog's comments collection
        blog.addComment(savedComment);
        
        return savedComment;
    }

    /**
     * Get comments by blog
     */
    @Transactional(readOnly = true)
    public List<Comment> getCommentsByBlog(Long blogId) {
        if (!blogRepository.existsById(blogId)) {
            throw new IllegalArgumentException("Blog not found with id: " + blogId);
        }
        return commentRepository.findByBlogIdOrderByCreatedAtAsc(blogId);
    }

    /**
     * Delete comment
     */
    public void deleteComment(Long commentId) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new IllegalArgumentException("Comment not found with id: " + commentId));
        
        // Remove from blog's comments collection
        Blog blog = comment.getBlog();
        if (blog != null) {
            blog.removeComment(comment);
        }
        
        commentRepository.delete(comment);
    }

    /**
     * Search blogs by title or content
     */
    @Transactional(readOnly = true)
    public List<Blog> searchBlogs(String searchTerm) {
        if (searchTerm == null || searchTerm.trim().isEmpty()) {
            return getAllPublishedBlogs();
        }
        return blogRepository.findByTitleContainingIgnoreCaseOrContentContainingIgnoreCaseAndPublishedTrue(
                searchTerm);
    }

    /**
     * Count total blogs
     */
    @Transactional(readOnly = true)
    public Long countAllBlogs() {
        return blogRepository.count();
    }

    /**
     * Count published blogs
     */
    @Transactional(readOnly = true)
    public Long countPublishedBlogs() {
        return blogRepository.countByPublishedTrue();
    }

    /**
     * Count blogs by author
     */
    @Transactional(readOnly = true)
    public Long countBlogsByAuthor(Long authorId) {
        return blogRepository.countByAuthorId(authorId);
    }
}
